WindowServer: Add support for default MenuItem

tomuta opened the pull request:
This allows marking a MenuItem as a default action, e.g. in a
context menu for an action that reflects what e.g. a double click
would perform.

Also enhance the window menu to mark the close action as the
default, and when double clicked perform that action.

I keep finding myself trying to close windows by double clicking on the window menu (I know, bad behavior)... This makes it more "natural" to my habits. No, I haven't used Windows much for a long, long time, but it's hard to get rid of habits...

![Screenshot from 2020-07-07 11-25-47](https://user-images.githubusercontent.com/10320822/86819293-ab272580-c044-11ea-967c-278c6cf08789.png)

awesomekling commented:
The concept of a default menu action doesn't seem generally useful outside of this one specific situation.
Where else would it be used?

tomuta commented:
> The concept of a default menu action doesn't seem generally useful outside of this one specific situation.
> Where else would it be used?

E.g. right clicking something in File Manager, the action that would be performed by a double click (open the folder, open the file, execute a program, etc)?

tomuta commented:
To me it's a usability feature I miss from the good old days ;-)

awesomekling commented:
Ohhhh the text is in *bold*! I didn't realize, duh. Now I get it! :smile:

awesomekling commented:
And by that I mean **bold**, obviously.

tomuta commented:
Yeah, it highlights what it would have done if you had clicked or double-clicked something (if there is a context menu).

tomuta commented:
![Screenshot from 2020-07-07 16-45-16](https://user-images.githubusercontent.com/10320822/86852394-ee4bbd80-c071-11ea-8cf4-d17a6b184592.png)

diversys commented:
This seems to fix https://github.com/SerenityOS/serenity/issues/1289.

bugaevc commented:
> E.g. right clicking something in File Manager, the action that would be performed by a double click (open the folder, open the file, execute a program, etc)?

So, an action that would be performed on activation?

Does anything ensure that this default action actually does the same thing as activation, or is it up to individual apps/widgets to ensure?

awesomekling commented:
> > E.g. right clicking something in File Manager, the action that would be performed by a double click (open the folder, open the file, execute a program, etc)?
> 
> So, an action that would be performed on activation?
> 
> Does anything ensure that this default action actually does the same thing as activation, or is it up to individual apps/widgets to ensure?

Great question. Can we find a way to make this programmatic?

Also, I don't think it makes sense for an action to be "default", as an action can occur in multiple context.
You can add the same action to a toolbar, menu, etc.

It seems that `is_default` should be a parameter to the `add_action()` API.

tomuta commented:
> > > E.g. right clicking something in File Manager, the action that would be performed by a double click (open the folder, open the file, execute a program, etc)?
> > 
> > 
> > So, an action that would be performed on activation?
> > Does anything ensure that this default action actually does the same thing as activation, or is it up to individual apps/widgets to ensure?
> 
> Great question. Can we find a way to make this programmatic?
> 
> Also, I don't think it makes sense for an action to be "default", as an action can occur in multiple context.
> You can add the same action to a toolbar, menu, etc.

It's only a visual indicator (bold text) applied to menu items. Just like it is only a visual indicator on menu items. Similar to it e.g. being checked or not.

> It seems that `is_default` should be a parameter to the `add_action()` API.
I thought about that, but since this is a rather rarely used feature I thought it would be better to just add a separate `set_default` call if one really wants an action marked as "default".

Maybe we just need to come up with a better name? If I remember correctly from my old Windows days, that's what they call it there, and similarly it's just a visual indicator.

awesomekling commented:
My point is that the action may be the default behavior in one context, but not in another.

tomuta commented:
> My point is that the action may be the default behavior in one context, but not in another.

Hm. What if we when realizing a context menu provide an optional default action, and if then compares the actions in the menu to that, and if it matches then it calls `MenuItem::set_default`? This way, actions wouldn't have a "default" flag, but when you display a context menu then you could also indicate what you consider the "default", and the menu would get populated as expected?

tomuta commented:
How does this look? I reworked it, so that now you can specify an optional default action when calling `Menu::popup`, since these things are clearly only applicable to context menu items anyway. I removed all the default stuff from `Action`. I quite like this ;)

awesomekling commented on Libraries/LibGUI/Menu.h:
```diff
@@ -56,19 +56,20 @@ class Menu final : public Core::Object {
     int m_menu_id { -1 };
     String m_name;
     NonnullOwnPtrVector<MenuItem> m_items;
+    Action* m_last_default_action { nullptr };
```
Using a raw pointer like this is pretty sloppy. Let's use a `WeakPtr<Action>` instead :)

tomuta replied:
Well, the pointer is only ever compared to determine a match, the object is never accessed using it. Still want me to change it to a WeakPtr instead?

tomuta replied:
Ok, changed it to a WeakPtr

nico commented:
If I right-click a `.font` file in File Manager, the context menu shows in bold "Open in Text Editor...", but if I double-click the file it opens in Font Editor. is this expected?

tomuta commented:
> If I right-click a `.font` file in File Manager, the context menu shows in bold "Open in Text Editor...", but if I double-click the file it opens in Font Editor. Is this expected?

That was just a missing feature. Added a pull request: #2795
